### ISSUE TITLE
Issue with starting a FL process with a trained model

### ISSUE THREAD

**Issue by zyzhang1130** [alibaba/FederatedScope#477]
So I was trying to continue training a model with vanilla FL by copying the model into the server and client models as follows:
```
Fed_runner = FedRunner(data=data,
                               server_class=get_server_cls(cfg),
                               client_class=get_client_cls(cfg),
                               config=cfg.clone())
        
Fed_runner.server.model = copy.deepcopy(pretrained_model)
for i in range(cfg.federate.client_num):
       Fed_runner.client[i+1].model = copy.deepcopy(pretrained_model)
Fed_runner.run()
```

I was expecting the training accuracy to continue to rise as `pretrained_model` was trained on the same data but not until convergence. but the training accuracy only fluctuated around the same level. Is this because I used the wrong way to copy the model weights?

**Comment by zyzhang1130** [comment]
> So I was trying to continue training a model with vanilla FL by copying the model into the server and client models as follows:
> 
> ```
> Fed_runner = FedRunner(data=data,
>                                server_class=get_server_cls(cfg),
>                                client_class=get_client_cls(cfg),
>                                config=cfg.clone())
>         
> Fed_runner.server.model = copy.deepcopy(pretrained_model)
> for i in range(cfg.federate.client_num):
>        Fed_runner.client[i+1].model = copy.deepcopy(pretrained_model)
> Fed_runner.run()
> ```
> 
> I was expecting the training accuracy to continue to rise as `pretrained_model` was trained on the same data but not until convergence. but the training accuracy only fluctuated around the same level. Is this because I used the wrong way to copy the model weights?

I changed it to the following:
```
Fed_runner = FedRunner(data=data,
                               server_class=get_server_cls(cfg),
                               client_class=get_client_cls(cfg),
                               config=cfg.clone())
        
Fed_runner.server.model.load_state_dict(pretrained_model.state_dict().copy())

Fed_runner.run()
```
and it seemed the issue disappeared. I am still curious to know why `deepcopy` didn't work here though, as I thought it worked in the same way as `load_state_dict` in most of the cases.

**Comment by xieyxclack** [comment]
Thank you for your interest and sorry for the late response!  I would try to conduct some experiments to analyze what happens here

**Comment by xieyxclack** [comment]
@zyzhang1130 Hi, for reproducing, could you tell me how to instantiate `pretrained_model`? By `get_model` or copying from server/client' model and loading saved checkpoint via `torch.load`?

**Comment by zyzhang1130** [comment]
Is `get_model` a built-in way to load a pre-trained model (didn't find it in configs)? As mentioned I tried deep copy which didn't work, but loading state dictionary did work so I stick with that for now.

**Comment by xieyxclack** [comment]
> Is `get_model` a built-in way to load a pre-trained model (didn't find it in configs)? As mentioned I tried deep copy which didn't work, but loading state dictionary did work so I stick with that for now.

Could you provide the source code related to the instantiation of `pretrained_model`  mentioned above?

**Comment by zyzhang1130** [comment]
this is what I can find about [pytorch state dictionary](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.load_state_dict)
this is [deepcopy's source code](https://github.com/python/cpython/blob/3.11/Lib/copy.py )

**Comment by xieyxclack** [comment]
> > So I was trying to continue training a model with vanilla FL by copying the model into the server and client models as follows:
> > ```
> > Fed_runner = FedRunner(data=data,
> >                                server_class=get_server_cls(cfg),
> >                                client_class=get_client_cls(cfg),
> >                                config=cfg.clone())
> >         
> > Fed_runner.server.model = copy.deepcopy(pretrained_model)
> > for i in range(cfg.federate.client_num):
> >        Fed_runner.client[i+1].model = copy.deepcopy(pretrained_model)
> > Fed_runner.run()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I was expecting the training accuracy to continue to rise as `pretrained_model` was trained on the same data but not until convergence. but the training accuracy only fluctuated around the same level. Is this because I used the wrong way to copy the model weights?
> 
> I changed it to the following:
> 
> ```
> Fed_runner = FedRunner(data=data,
>                                server_class=get_server_cls(cfg),
>                                client_class=get_client_cls(cfg),
>                                config=cfg.clone())
>         
> Fed_runner.server.model.load_state_dict(pretrained_model.state_dict().copy())
> 
> Fed_runner.run()
> ```
> 
> and it seemed the issue disappeared. I am still curious to know why `deepcopy` didn't work here though, as I thought it worked in the same way as `load_state_dict` in most of the cases.

Hi, I think you misunderstand what I ask for. I want to know how you create the `pretrained_model` in these provided codes (such as `pretrained_model = xxx`) and I want to reproduce the results first. Thank you :)

**Comment by zyzhang1130** [comment]
I see. It was simply taken from the previous FL training. Example: after finishing one Fed_runner.run() from scratch, let `pretrained_model = Fed_runner.server.model`

**Comment by xieyxclack** [comment]
I have re-produced the mentioned issue, and found the reason:
- `Fed_runner.server.model = copy.deepcopy(pretrained_model)` will break synchronization between `server.model` and `server.models[0]` (refer to different objects) when `model_num=1`. And in the FS course,  server would maintain `server.models` (see [here](https://github.com/alibaba/FederatedScope/blob/c6a7de482047a4c84224fbc2744aed30811f4213/federatedscope/core/workers/server.py#L474)) but broadcast the model param obtained from `server.model` (see [here](https://github.com/alibaba/FederatedScope/blob/c6a7de482047a4c84224fbc2744aed30811f4213/federatedscope/core/workers/server.py#L666)). And using `Fed_runner.server.model.load_state_dict(pretrained_model.state_dict().copy())` would not break such synchronization.
- You can try to change `Fed_runner.server.model = copy.deepcopy(pretrained_model)` to `Fed_runner.server.models = [copy.deepcopy(pretrained_model)]` and observer the FL course can run as expected.
- To avoid such issue, the suggested way to save/restore the pretrained model is specify `federate.save_to` and `federate.restore_from`, more details can be found in the [docs](https://github.com/alibaba/FederatedScope/tree/master/federatedscope/core/configs#federate-general-fl-setting).


Also, shall we change the behavior [here](https://github.com/alibaba/FederatedScope/blob/master/federatedscope/core/workers/server.py#L667)? For example, `model_para = {} if skip_broadcast else self.models[0].state_dict()` @rayrayraykk

**Comment by zyzhang1130** [comment]
Thanks a lot for the clarification. As saving and loading model is an intermediate step in my pipeline (there are many models to be used as initializations), I didn't bother to save them one by one and stored them in a list instead.  So long as the aforementioned method or copying state dictionary worked all is well.
